internal/services/v0/namespace.go

```diff
@@ -43,119 +42,118 @@ func NewNamespaceServer() v0.NamespaceServiceServer {
 
 func (nss *nsServer) WriteConfig(ctx context.Context, req *v0.WriteConfigRequest) (*v0.WriteConfigResponse, error) {
 	ds := datastoremw.MustFromContext(ctx)
-	nsm := namespace.NewNonCachingNamespaceManager()
 
-	readRevision, _ := consistency.MustRevisionFromContext(ctx)
-
-	for _, config := range req.Configs {
-		// Validate the type system for the updated namespace.
-		ts, terr := namespace.BuildNamespaceTypeSystemWithFallback(core.ToCoreNamespaceDefinition(config), nsm, core.ToCoreNamespaceDefinitions(req.Configs), readRevision)
-		if terr != nil {
-			return nil, rewriteNamespaceError(ctx, terr)
-		}
-
-		vts, tverr := ts.Validate(ctx)
-		if tverr != nil {
-			return nil, rewriteNamespaceError(ctx, tverr)
-		}
-
-		if err := namespace.AnnotateNamespace(vts); err != nil {
-			return nil, rewriteNamespaceError(ctx, err)
-		}
+	revision, err := ds.ReadWriteTx(ctx, func(ctx context.Context, rwt datastore.ReadWriteTransaction) error {
+		for _, config := range req.Configs {
+			// Validate the type system for the updated namespace.
+			ts, terr := namespace.BuildNamespaceTypeSystemWithFallback(
+				core.ToCoreNamespaceDefinition(config),
+				rwt,
+				core.ToCoreNamespaceDefinitions(req.Configs),
+			)
+			if terr != nil {
+				return terr
+			}
 
-		// Ensure that the updated namespace does not break the existing tuple data.
-		//
-		// NOTE: We use the datastore here to read the namespace, rather than the namespace manager,
-		// to ensure there is no caching being used.
-		existing, _, err := ds.ReadNamespace(ctx, config.Name, readRevision)
-		if err != nil && !errors.As(err, &datastore.ErrNamespaceNotFound{}) {
-			return nil, rewriteNamespaceError(ctx, err)
-		}
+			vts, tverr := ts.Validate(ctx)
+			if tverr != nil {
+				return tverr
+			}
 
-		diff, err := namespace.DiffNamespaces(existing, core.ToCoreNamespaceDefinition(config))
-		if err != nil {
-			return nil, rewriteNamespaceError(ctx, err)
-		}
+			if err := namespace.AnnotateNamespace(vts); err != nil {
+				return err
+			}
 
-		headRevision, _ := consistency.MustRevisionFromContext(ctx)
-
-		for _, delta := range diff.Deltas() {
-			switch delta.Type {
-			case namespace.RemovedRelation:
-				qy, qyErr := ds.QueryTuples(
-					ctx,
-					&v1.RelationshipFilter{
-						ResourceType:     config.Name,
-						OptionalRelation: delta.RelationName,
-					},
-					headRevision,
-					options.WithLimit(options.LimitOne),
-				)
-				err = shared.ErrorIfTupleIteratorReturnsTuples(
-					ctx,
-					qy,
-					qyErr,
-					"cannot delete relation `%s` in definition `%s`, as a relationship exists under it", delta.RelationName, config.Name)
-				if err != nil {
-					return nil, rewriteNamespaceError(ctx, err)
-				}
+			// Ensure that the updated namespace does not break the existing tuple data.
+			existing, _, err := rwt.ReadNamespace(ctx, config.Name)
```

Comment:
I don't think we need this anymore, since we're inside a transaction. We might be able to just grab the namespace from the validated type system, which has it in memory

Reply:
I don't want to spend any time on v0 if I don't have to. Is it correct as written?

Reply:
Yeah, its just extra work for now, but that's not any different than now